### PR TITLE
Remove redundant copies (and some minor modifications)

### PR DESCRIPTION
**Pull request by munehiro-k**
Hi! Thank you for sharing your handy and easy-to-employ library.
I looked into the code and found some lines that might be able to improve mainly in efficiency, so I am making this pull request.

My modifications are as follows.
1. `ifstream` instance may be better initialized with `std::ios::binary` mode.
   * This fix is similar to your commit 7fe5fb8.
2. Some pass-by-value function calls cause redundant copies.
3. We can cut more copies with a move constructor and a move assignment operator.
4. I wanted a member function of `NDArray` that returns the underlying vector container.
   * My use case is to load a .npy file that contains 1d time series data and to perform a customized moving window operation, so plain `std::vector` and your load and save functionality of `NDArray` suffice my requirements.

I am afraid I have not fully tested all of my modifications, and I just made a simple code and checked that memory usage reduces compared to the original code.
If you share me your test codes, then I will perform the tests to check my modifications.

Feel free to respond conservatively. Anyway I appreciate your work, and I am able to proceed my task with your library, so thanks again!

Lastly, I paste my simple code to check the memory usage, just for your infomation.
```
#include <string>
#include <vector>
#include <iostream>

#include "ndarray.hpp"

int main() {
    // CATION
    // This code writes a barren file of size 2GB to your disk!!

    const std::string npy_path("some_big_array.npy");
    {
        const unsigned dim1 = 1024;
        const unsigned dim2 = 1024;
        const unsigned dim3 = 500;
        const unsigned size = dim1 * dim2 * dim3;

        std::vector<float> data(size, 0);

        // constructor with copy for a large vector
        // NDArray<float> array1(data,  {dim1, dim2, dim3});

        // constructor with move semantics for a large vector
        NDArray<float> array1(std::move(data), {dim1, dim2, dim3});

        std::cout << "ndarray initialized:\n";
        for(unsigned i=0; i<10; i++){
            std::cout << array1[i] << std::endl;
        }

        array1.save(npy_path);
        std::cout << "ndarray saved" << std::endl;
    }

    // move constructor 
    NDArray<float> array2 = NDArray<float>::load(npy_path);

    // copy constructor
    // NDArray<float> array3(array2);

    std::cout << "ndarray loaded again:\n";
    for(unsigned i=0; i<10; i++){
        std::cout << array2[i] << std::endl;
    }

    return 0;
}
```

### CODE REVIEW

File: include/ndarray.hpp [hunk 1]
```diff
@@ -94,8 +98,11 @@ class NDArray {
   //==========================================================================
   // Constant Methods
 
+  // Return underlying data vector
+  const std::vector<T>& data() const;
```
Comment (HunterBelanger) [inline include/ndarray.hpp]: I think having a way to return a reference to the underlying container is a great idea, though I think a different name should be used. When I see `data`, I would expect to get a pointer back, like with the STL containers. Perhaps we could rename this to be `data_vector` ? If you have another idea, I am very open to suggestions. After I merge your PR, I will probably add a `data` method which returns a pointer to the data to make it a little more STL-like.

File: include/ndarray.hpp [hunk 2]
```diff
@@ -110,20 +117,24 @@ class NDArray {
...
-  void reshape(std::vector<size_t> new_shape);
+  void reshape(const std::vector<size_t>& new_shape);
 
   // Realocates array to fit the new size.
   // DATA CAN BE LOST IF ARRAY IS SHRUNK
-  void reallocate(std::vector<size_t> new_shape);
+  void reallocate(const std::vector<size_t>& new_shape);
+
+  // Assignment Operator
+  NDArray& operator=(const NDArray&) = default;
```
Comment (HunterBelanger) [inline include/ndarray.hpp]: These are fine, but could they just be moved to the top of the class, so they are with the constructors ?

File: include/ndarray.hpp [hunk 3]
```diff
@@ -436,7 +477,12 @@ NDARRAY_INLINE const T& NDArray<T>::operator[](size_t i) const {
 }
 
 template <class T>
-NDARRAY_INLINE std::vector<size_t> NDArray<T>::shape() const {
+NDARRAY_INLINE const std::vector<T>& NDArray<T>::data() const {
```
Comment (HunterBelanger) [inline include/ndarray.hpp]: This is just a reminder that the name of the method will need to be changed here as well.

**Review by HunterBelanger — COMMENTED** [review]
Hello ! I am glad to hear that you have found this little tool useful, and thank you so much for submitting these performance improvements ! I wrote this when I was still somewhat new to C++, and definitely wasn't paying attention to all the details. I left a few comments about some minor changes, but other than those, I think this looks good. I must admit that I don't have an actual test suite for this tool. I will try using it in one of my programs which uses it, and check if I encounter any unexpected results.

**Comment by munehiro-k** [comment]
Thank you for reviewing my PR! I fixed my commits to fit your comments.

Not enough to replace a test suite, my code to perform a moving window operation that uses `NDArray` to load and save .npy files is working as intended with commit bf392fc.

File: include/ndarray.hpp (hunk 1 above)
Comment (munehiro-k) [reply include/ndarray.hpp]: The name `data_vector` seems just fine. I have changed my commit from 4ffd26e to bf392fc.
diff: https://github.com/munehiro-k/ndarray/compare/4ffd26e..bf392fc

The idea to make `data` member function to return a pointer sounds also good. I have missed to consider to be STL-like.

File: include/ndarray.hpp (hunk 2 above)
Comment (munehiro-k) [reply include/ndarray.hpp]: Your comment is just right. I have moved the declaration of assignment operators near constructors with former commit aef7e48 and updated commit b7886e2.
diff: https://github.com/munehiro-k/ndarray/compare/aef7e48..b7886e2

File: include/ndarray.hpp (hunk 3 above)
Comment (munehiro-k) [reply include/ndarray.hpp]: Thanks for the kind reminder.

**Comment by HunterBelanger** [comment]
Thanks for those changes and thank you for your great contributions as well ! It all looks good to me, so I am going to go ahead and merge this in.

**Pull request merged**